code/modules/mod/mod_modules.dm

```diff
@@ -0,0 +1,616 @@
+	animate(mod.wearer, alpha = 255, time = 1.5 SECONDS)
+
+/obj/item/mod/module/stealth/proc/unstealth(datum/source)
+	SIGNAL_HANDLER
+
+	to_chat(mod.wearer, "<span class='warning'>[src] gets discharged from contact!</span>")
+	do_sparks(2, TRUE, src)
+	drain_power(use_power_cost)
+	on_deactivation()
+
+/obj/item/mod/module/stealth/proc/on_unarmed_attack(datum/source, atom/target)
+	SIGNAL_HANDLER
+
+	if(!isliving(target))
+		return
+	unstealth(source)
+
+/obj/item/mod/module/stealth/proc/on_bullet_act(datum/source, obj/projectile/projectile)
+	SIGNAL_HANDLER
+
+	if(!projectile.nodamage)
+		unstealth(source)
+
+/obj/item/mod/module/stealth/ninja
+	name = "MOD advanced cloaking module"
+	desc = "A module using advanced cloaking technology to hide the user from plain sight."
+	bumpoff = FALSE
+	stealth_alpha = 20
+	active_power_cost = 10
+	use_power_cost = 50
+	cooldown_time = 3 SECONDS
+
+/obj/item/mod/module/jetpack
+	name = "MOD ion jetpack module"
+	desc = "A module that runs a micro-jetpack using a MOD's power cell."
+	module_type = MODULE_TOGGLE
+	complexity = 3
+	active_power_cost = 20
+	use_power_cost = 80
+	incompatible_modules = list(/obj/item/mod/module/jetpack)
+	cooldown_time = 0.5 SECONDS
+	var/full_speed = FALSE
+	var/datum/effect_system/trail_follow/ion/ion_trail
+
+/obj/item/mod/module/jetpack/Initialize()
+	. = ..()
+	ion_trail = new
+	ion_trail.auto_process = FALSE
+	ion_trail.set_up(src)
+
+/obj/item/mod/module/jetpack/Destroy()
+	QDEL_NULL(ion_trail)
+	return ..()
+
+/obj/item/mod/module/jetpack/on_activation()
+	. = ..()
+	if(!. || !allow_thrust())
+		return
+	ion_trail.start()
+	RegisterSignal(mod.wearer, COMSIG_MOVABLE_MOVED, .proc/move_react)
+	RegisterSignal(mod.wearer, COMSIG_MOVABLE_PRE_MOVE, .proc/pre_move_react)
+	RegisterSignal(mod.wearer, COMSIG_MOVABLE_SPACEMOVE, .proc/spacemove_react)
+	if(full_speed)
+		mod.wearer.add_movespeed_modifier(/datum/movespeed_modifier/jetpack/fullspeed)
+
+/obj/item/mod/module/jetpack/on_deactivation(mob/user)
+	. = ..()
+	if(!.)
+		return
+	ion_trail.stop()
+	UnregisterSignal(mod.wearer, COMSIG_MOVABLE_MOVED)
+	UnregisterSignal(mod.wearer, COMSIG_MOVABLE_PRE_MOVE)
+	UnregisterSignal(mod.wearer, COMSIG_MOVABLE_SPACEMOVE)
+	mod.wearer.remove_movespeed_modifier(/datum/movespeed_modifier/jetpack/fullspeed)
+
+/obj/item/mod/module/jetpack/proc/move_react(mob/user)
+	if(!active)//If jet dont work, it dont work
+		return
+	if(!mod.wearer?.client)//Don't allow jet self using
+		return
+	if(!isturf(mod.wearer.loc))//You can't use jet in nowhere or from mecha/closet
+		return
+	if(!(mod.wearer.movement_type & FLOATING) || mod.wearer.buckled)//You don't want use jet in gravity or while buckled.
+		return
+	if(mod.wearer.pulledby)//You don't must use jet if someone pull you
+		return
+	if(mod.wearer.throwing)//You don't must use jet if you thrown
+		return
+	if(length(mod.wearer.client.keys_held & mod.wearer.client.movement_keys))//You use jet when press keys. yes.
```

Comment:
what if an AI is the one in control?

Reply:
also, this means that my modsuit-wearing NPC monkey army can't jetpack around in space, and that makes me sad